Adding capability for charged particle motion in electric and magnetic fields (#20)
idaholab/salamander: doc/content/source/userobjects/ParticleStepperBase.md

# ParticleStepperBase

> `ParticleStepperBase` is the abstract class form which all steppers should be derived. `ParticleStepperBase` provides the basic method to update a particles direction and maximum distance based on the velocity stored in ray data, the current time step and dimension of the mesh. In the provided method `setMaxDistanceAndDirection` the direction is set based on the dimension of the mesh, and the maximum distance is set with the following rule:

Reviewer: ```suggestion
`ParticleStepperBase` is the abstract class form which all steppers should be derived. `ParticleStepperBase` provides the basic method to update a particle's direction and maximum distance based on the velocity stored in ray data, the current time step, and dimension of the mesh. In the provided method `setMaxDistanceAndDirection` the direction is set based on the dimension of the mesh, and the maximum distance is set with the following rule:
```

\begin{equation}
  d_\text{max} = \vec{v} \cdot \vec{v} \: dt,
\end{equation}

> Where $d_\text{max}$ is maximum distance the particle is allowed to travel during the nex time step, and $\vec{v}$ is the particles' velocity.

Reviewer: ```suggestion
where $d_\text{max}$ is maximum distance the particle is allowed to travel during the next time step, and $\vec{v}$ is the particles' velocity.
```

# ParticleStepperBase

> `ParticleStepperBase` is the abstract class form which all steppers should be derrived. `ParticleStepperBase` provides the basic method to update a particles direction and maximum distance based on the veloctiy stored in ray data, the current time step and dimension of the mesh. Addtionally, several other useful methods are provided. In the provided method `setMaxDistanceAndDirection` the particles velocity is updated to the newly calculated velocity, the direction is set based on the dimension of the mesh, and the maximum distance is set with the following rule.

Reviewer: ```suggestion
`ParticleStepperBase` is the abstract class form which all steppers should be derived. `ParticleStepperBase` provides the basic method to update a particles direction and maximum distance based on the velocity stored in ray data, the current time step and dimension of the mesh. Addtionally, several other useful methods are provided. In the provided method `setMaxDistanceAndDirection` the particles velocity is updated to the newly calculated velocity, the direction is set based on the dimension of the mesh, and the maximum distance is set with the following rule:
```
Could you expand and provide more details about `Addtionally, several other useful methods are provided`?

Author: I have removed this and now it only discusses the other method that is still

\begin{equation}
  d_\text{max} = \vec{v} \cdot \vec{v} \: dt

> \end{equation}

Reviewer: ```suggestion
\begin{equation}
  d_\text{max} = \vec{v} \cdot \vec{v} \: dt,
\end{equation}
with ....
```
Please explain the notations.

> However, only the components of the velocity data corrisponding to the dimension of the mesh will be used in the calculation. If the mesh is 2D then only the first 2 components of velocity will be used for this calculation. This calculation is peformed with `ParticleStepperBase::setupStep`. This function takes an argument of the ray that needs its velocity updated, the new velocity of the ray and a time step to be used for the maximum distance calculation to allow flexibility for particle pushing methods.

Reviewer: ```suggestion
However, only the components of the velocity data corresponding to the dimension of the mesh will be used in the calculation. If the mesh is 2D then only the first 2 components of velocity will be used for this calculation. This calculation is performed with `ParticleStepperBase::setupStep`. This function takes an argument of the ray that needs its velocity updated, the new velocity of the ray and a time step to be used for the maximum distance calculation to allow flexibility for particle pushing methods.

!syntax parameters /UserObjects/ParticleStepperBase

!syntax inputs /UserObjects/ParticleStepperBase

!syntax children /UserObjects/ParticleStepperBase
```

Author: The !syntax commands on this page have been excluded. The particle stepper base is an abstract base class and will never be used directly by end users. These also break the docs when they are added.
